modules/admin_manual/pages/configuration/server/security/hsmdaemon.adoc

```diff
@@ -0,0 +1,588 @@
+:jwt-url: https://jwt.io/
+:network-sockets-url: https://en.wikipedia.org/wiki/Network_socket
+:go-install-url: https://golang.org/doc/install
+:opensc-wiki-url: https://github.com/OpenSC/OpenSC/wiki
+:opensuse-security-repositories-url: https://download.opensuse.org/repositories/security/
+:php-exec-function-url: https://www.php.net/manual/en/function.exec.php
+:pkcs11-url: https://en.wikipedia.org/wiki/PKCS_11
+:pkcs11-tool-url: https://linux.die.net/man/1/pkcs11-tool 
+:softhsm2-url: https://www.opendnssec.org/softhsm/
+:unix-sockets-url: http://beej.us/guide/bgipc/html/multi/unixsock.html
+
+The hsmdaemon is a daemon to delegate encryption to an {hsm-url}[HSM (Hardware Security Module)].
+It can be necessary, as PHP cannot directly interface with {pkcs11-url}[a PKCS11 stack]; neither by an API wrapper, because one does not exist, yet, nor via the OpenSSL bindings
+Because of this, a separate process is needed to decrypt anything with the private key stored in an HSM. 
+
+Running {php-exec-function-url}[exec()] to decrypt the key with a command line command to do the encryption might leak the HSM credentials if the admin lists the currently running processes. 
+To prevent that an "HSM" daemon will be used that can open a session to the HSM upon startup. 
+
+This daemon will be used by ownCloud to decrypt the current master key upon request. 
+The communication happens via {unix-sockets-url}[UNIX sockets] or {network-sockets-url}[TCP sockets] and is authorized by a shared token that the daemon stores in the ownCloud database via a REST/JSON route.
+
+ownCloud internally uses OpenSSL to en-/decrypt keys and needs to be extended to support en-/decrypt operations via the new daemon. 
+The current solution would be to encrypt the current ownCloud master key with a key from the HSM. 
+
+== How The HSM Daemon Interacts with ownCloud
+
+Upon startup, the daemon will generate a token and send it to ownCloud via a new REST/JSON route. 
+After connecting with the HSM daemon, an unsophisticated, line-based, protocol is used (every line ends with CRLF):
+
+. ownCloud sends the token read from database.
+. The daemon compares the received token with its token and returns an `"OK"` line.
+. ownCloud then sends the data it wants to decrypt as a {base64-encoding-url}[Base64-encoded], one-line string.
+. The daemon returns the decrypted data as a Base64-encoded one-line string.
+
+Doing so ensures that an evil admin will need to wiretap the communication between either the database or the HSM daemon and ownCloud.
+
+== Quick Overview
+
+HSM support consists of two core parts:
+
+. An actual HSM PKCS11 module.
+. An hsmdaemon that provides a {jwt-url}[JWT]-protected web API for the PKCS11 stack to generate key pairs and decrypt data.
+
+== Deployment Recommendation
+
+We recommend running hsmdaemon on every Apache server to reduce latency. 
+
+== Installation
+
+Integrating the hsmdaemon with ownCloud requires X steps; these are:
```

Comment:
`requires X steps` ?
`requires some steps`

Reply:
Well spotted. I need to add some kind of generic marker to remind me when I do this.